### ISSUE TITLE
Correlation engine on fuzzy hashing or levenshtein distance

### ISSUE THREAD

**Issue by adulau** [MISP/MISP#1279]
In the future version of MISP, it would be nice to have a correlation engine on fuzzy hashing allowing
to correlate information which is sharing some similarities.

The example is the following. In the financial sector, you might have name of bank account holders which are reusing same stolen documents but the name are not exactly the same. As this might be difficult to share the full name, the idea was to share an "ssdeep-like" (context triggered piecewise hashes - CTPH) hash for the string and to make the correlation on a factor of similarities.  The concept can be reused by many others use-cases where you have typographic errors in names, string or even numbers.

**Comment by adulau** [comment]
Maybe at a first step, Levenshtein distance for specific attribute types is more realistic to do.

A simple task crawling all the same attributes type, calculating the levenshtein distance and added it in a ZRANK for each attributes. Then the user can tune the value of the ranked set as they wish.,

**Comment by plutec** [comment]
I'm working in a fuzzy "database" to find fuzzy hashes (ssdeep mainly, but portable to other hashes) using a similarity function (defined by the user). It's planned to publish it in the next weeks. 
The first version is programmed in python, but it is portable to another programming language without problems.

Maybe that will be helpful for the project.

**Comment by plutec** [comment]
FYI, we will use these for malware correlation.

**Comment by adulau** [comment]
One of the idea is to find typo entered in MISP attributes like incorrect hashes, missed names or strings.

We will implement it as a MISP module to allow users to find fuzzy matching strings that are close to a specific threshold. We did some tests and it seems that the normalized version of the levenshtein algorithm gives good results.

``` python
import distance
t2 = ("db405ad775ac887a337b02ea8b07fdcd")
t1 = ("db405ad775ac887a337b02ea8b07fddc")
distance.nlevenshtein(t1,t2)

Out[49]: 0.0625
```

For bigger value (e.g. binaries and samples), we already did some work (https://github.com/CIRCL/ssdc/tree/master/multiproc and https://github.com/MISP/misp-workbench/tree/master/grouping) using ssdeep on this aspect but it's not in the core MISP until now.

Maybe we could integrate your approach as a misp-module too. Thanks a lot.
